### DISCUSSION TITLE
ViT not learning for binary classification problem

### DISCUSSION THREAD

**Discussion by bax24** [Project-MONAI/MONAI#4129] [Q&A]
Hello everyone !

I didn't know whether to ask this in the discussion forum or to start a "New issue" so I hope this is the right place..

I am currently trying to implement a classification network with MRI scans (from ADNI) to detect alzheimer. I already implemented 3D CNN network which works quite well and I wanted to test the backbone of UNETR, so the ViT, to see what kind of results I could get.

However, the network doesn't seem to learn as the loss remains the same throughout epochs. I tried with a classification of 1 class and 2 classes and the results are the same. I suppose I am not using it correctly and I was wondering if you could help me ?

<img width="751" alt="main" src="https://user-images.githubusercontent.com/38333245/163220818-f375a947-5420-4373-86cc-b42baf8d7164.png">

<img width="507" alt="model" src="https://user-images.githubusercontent.com/38333245/163220779-d413ce2c-b015-4b04-9109-2baf2ac63cca.png">

<img width="852" alt="train" src="https://user-images.githubusercontent.com/38333245/163220854-93cb1b6e-c895-49e9-b482-94ae10eb95b1.png">

<img width="885" alt="test" src="https://user-images.githubusercontent.com/38333245/163220888-27232af3-7e6d-44c5-94fa-1804ec88d230.png">

I attached the codes for my main, my model, how I train and test it. I hope this is ok I didn't know whether a github would be easier for you or not (since my code is still a bit messy). 

I hope to hear from you and I thank you in advance,

Kind regards,

Lucas

**Comment by finalelement** [comment]
Hi @bax24 thanks for starting the discussion. Could you share a github link with the CNN code as well? 

The ViT usage at the first glance does not seem to be incorrect, there does seem to be some confusion between out_channel and num_classes as that is not being tied together, please check and verify that.

Another thing that is interesting here is that how are you testing with a fixed input patch size. Is your CNN network also monai based?

**Comment by bax24** [comment]
Hello,

What do you mean when you say 'tied together' i'm sorry I don't get it? The sizes of my MRI is (96, 96, 48) and it only has one channel (I believe).

And for the patch size I just used (16, 16, 16) for training and testing, should I use something else ?

The CNN is not MONAI based, its implementation uses simple 3D CNN layers and fully connected ones.

Here is my github with the relevant codes : https://github.com/bax24/Alzheimer-Classification

Thanks again for your help!

PS : I am still a student and this is part of my master thesis so if you see something outrageous in the way that I am training or testing my models don't hesitate to tell me

**Reply by finalelement** [reply]
What I mean to say by 'not tied together' is here you define out_channel=1 [L204](https://github.com/bax24/Alzheimer-Classification/blob/874b60dd0ac50ac4f6e4e91b74a4c17ca8e81ff1/main.py#L204) but the out_channel is not being assigned to num_classes  [here]
(https://github.com/bax24/Alzheimer-Classification/blob/874b60dd0ac50ac4f6e4e91b74a4c17ca8e81ff1/model.py#L138)

I realize that you are downsampling all your images to be of the same resolution, with 96x96x48

For the patch size of ViT 16x16x16 is good and should work fine.

Can you share what you see with the CNN, the plots?

**Reply by bax24** [reply]
![E_CNN200-25-A(1)](https://user-images.githubusercontent.com/38333245/174811858-023af360-f7d9-41cb-9799-7e160641f95e.png)

Here is the plot you asked for i am sorry i forgot to send it

**Comment by bax24** [comment]
The CNN models now gives very good results but I still cannot make the ViT model learn (the loss is constant throughout the learning

No matter the epoch I get the following

![Capture d’écran 2022-06-18 à 17 19 31](https://user-images.githubusercontent.com/38333245/174445032-7ec13f81-5d1e-4d8e-8e4c-cfde77144807.png)

I have updated my github (https://github.com/bax24/Alzheimer-Classification) could you please take a look and tell me if i am doing something wrong ? Maybe with the backpropagation ?

Thanks a lot in advance,

Lucas

**Reply by finalelement** [reply]
I took a look at your code, here are suggestions which could help:

- https://github.com/bax24/Alzheimer-Classification/blob/ecd3aafc0e67c3e5a07f0e678224ae6803d009fd/model.py#L159 why is the activation not being used for the ViT part?
- Could try this with a lowered learning rate of probably 1e-4.
- I noticed the plots of training and testing/validation for the CNN model, looks quite nice, can you share these plots when you run ViT instead of CNN?

**Reply by bax24** [reply]
Also, should I rewrite (like in NetworkB in my github) the init and forward method or should I just use monai.Vit as the model itself ?

**Comment by bax24** [comment]
Here is the plot when running the ViT with a lower learning rate of 1e-4.

![ViT100-25-A-Norm_Roismalr](https://user-images.githubusercontent.com/38333245/175298559-1a852eb8-51c0-44a9-9038-c9b6292a550d.png)

Everything remain constant..

I do not use the activation because I already use a Sigmoid and it didn't work either with the log_softmax do you think I should try again ?

**Comment by chris-clem** [comment]
Hi,

I noticed that the Vision Transformer (ViT) with `classification=True` has `nn.Tanh()` in the end of the `classification_head` (https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/vit.py#L100). I did the following and it worked:

```python
model = monai.networks.nets.ViT(...)
model.classification_head = nn.Sequential(nn.Linear(768, num_classes))
```

Best
Christoph

**Reply by finalelement** [reply]
@bax24 the above is worth a shot, as you mentioned that you have an activation of sigmoid addiitionally.

**Reply by bax24** [reply]
When running the model proposed by @chris-clem I get the following error and I cannot figure it out

![Capture d’écran 2022-06-29 à 15 13 14](https://user-images.githubusercontent.com/38333245/176445028-56a6ac3b-8a4b-491d-a412-a331e67e24d9.png)

Which arguments are they talking about ?

**Reply by bax24** [reply]
I get the same mistakes when I run the example proposed on the doc.monai.io website :

" # for 3-channel with image size of (128,128,128), 24 layers and classification backbone
>>> net = ViT(in_channels=3, img_size=(128,128,128), pos_embed='conv', classification=True) "

After trying multiple solutions, the best I can get is the network varying for 3 or 4 epochs and then stays constant along all the rest

![ViT100-25-A-SC_Full-16-vit](https://user-images.githubusercontent.com/38333245/176996866-eae37bd7-8498-4835-bb7d-019d6b23dec2.png)

This plot is obtained with the updated github (https://github.com/bax24/Alzheimer-Classification)
